nlprule/src/rule/engine/composition.rs

```diff
@@ -284,50 +299,51 @@ pub struct OffsetAtom {
 }
 
 impl Atomable for OffsetAtom {
-    fn is_match(&self, input: &[Token], graph: &MatchGraph, position: usize) -> bool {
+    fn is_match(&self, context: Context, position: usize) -> bool {
+        let (sentence, _) = context;
         let new_position = position as isize + self.offset;
 
-        if new_position < 0 || (new_position as usize) >= input.len() {
+        if new_position < 0 || (new_position as usize) >= sentence.len() {
             false
         } else {
-            self.atom.is_match(input, graph, new_position as usize)
+            self.atom.is_match(context, new_position as usize)
         }
     }
 }
 
 #[derive(Debug, Default, Clone)]
 pub struct Group {
-    pub char_span: (usize, usize),
+    pub span: Span,
 }
 
 impl Group {
-    pub fn new(char_span: (usize, usize)) -> Self {
-        Group { char_span }
+    pub fn new(span: Span) -> Self {
+        Group { span }
     }
 
     pub fn tokens<'a, 't>(
         &'a self,
-        tokens: &'t [Token<'t>],
+        sentence: &'t MatchSentence,
     ) -> impl DoubleEndedIterator<Item = &'t Token<'t>> {
-        let start = self.char_span.0;
-        let end = self.char_span.1;
+        let start = self.span.char().start;
+        let end = self.span.char().end;
 
-        tokens.iter().filter(move |x| {
-            x.char_span.1 > x.char_span.0 // special tokens with zero range (e. g. SENT_START) can not be part of groups
-                    && x.char_span.0 >= start
-                    && x.char_span.1 <= end
+        sentence.iter().filter(move |x| {
+            x.span().char().end > x.span().char().start // special tokens with zero range (e. g. SENT_START) can not be part of groups
```

Comment:
Consider renaming `Span`s in a future PR. `Span`s in rust are commonly used in `rustc`/`syn::Span`s, which are `line` (starting at 1) and `column` (base 0) tuples. This could be applicable here, but I am not having a better name in mind besides `Range`.

Reply:
Good point! Do you know of any crate that has a struct named `Range` in the public API? I'd think it is still problematic because of confusion with `std::ops::Range`. 

I just found https://docs.rs/codespan/0.11.1/codespan/struct.Span.html which gives some precedent for naming a struct `Span` that is not a span in the `syn` sense so I think it's better to stick with that.

And as you mentioned if there is some clearly better name we can still of course switch to that in a future PR.

Reply:
I was thinking more about _reusing_ the `std::ops::Range` :)

Reply:
I've addressed in the comment below why I think that's problematic. The `Span` abstraction definitely makes sense internally, keeping track of two different implementations (one public, one internal) that do the same thing isn't worth it IMO.